### ISSUE TITLE
"Image elements have explicit `width` and `height`" audit ok even with percentage values

### ISSUE THREAD

**Issue by nhoizey** [GoogleChrome/lighthouse#14449]
### FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/main/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

### URL

https://jsbin.com/kadiwajeza

### What happened?

With this HTML (from linked jsbin):

```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>JS Bin</title>
</head>
<body>
  <img src="https://via.placeholder.com/300x200" width="300" height="200">
  <img src="https://via.placeholder.com/300x200">
  <img src="https://via.placeholder.com/300x200" width="100%" height="200">
</body>
</html>
```

Only the second image is reported as missing "explicit `width` and `height`":

![CleanShot 2022-10-14 at 16 01 34](https://user-images.githubusercontent.com/78213/195865892-f7816a50-c3da-412c-abd0-04eb0dc1e2dd.png)

Per MDN, [The intrinsic width of the image in pixels. Must be an integer without a unit](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Img#attr-width).

`width="100%"` is however accepted by browsers, they seem to use it as if it was `img { width: 100%; }` in CSS.

### What did you expect?

I would have expected Lighthouse to report `width="100%"` as not being an "explicit `width`".

### How were you running Lighthouse?

Chrome DevTools

### Lighthouse Version

9.6.2

### Chrome Version

105.0.5195.125 (Build officiel) (x86_64)

### OS

macOS

**Comment by adamraine** [comment]
> I would have expected Lighthouse to report width="100%" as not being an "explicit width".

`width: 100%` means the image element should fill 100% of the parent element's width. The parent element's width is known before the image downloads, so the size of the image element won't change once the image finishes downloading.

**Comment by nhoizey** [comment]
> `width: 100%` means the image element should fill 100% of the parent element's width

MDN states `%` should not be there. Is it allowed by the HTML spec?

> The parent element's width is known before the image downloads

If there is a `sizes` attribute, isn't it used to download the image ASAP, even before any rendering?

**Comment by eportis-cloudinary** [comment]
> the size of the image element won't change once the image finishes downloading

If the image's intrinsic aspect ratio is different from the parent element's width ÷ whatever's in the height attribute, it will. Also % seems to break the intrinsic aspect ratios for flexible images feature. https://codepen.io/eeeps/pen/BaxbLej?editors=1111

---

Some quick HTTP Archive querying says that only 1% of `<img>`s have "%" in their `width` (vs 31% of all [non-single-pixel]) `<img>`s.

MDN got their language from [the spec itself](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width), which states,

> The attributes, if specified, must have values that are [valid non-negative integers](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-non-negative-integer).

In addition,

> If both attributes are specified, then one of the following statements must be true:
>
> - specified width - 0.5 ≤ specified height * target ratio ≤ specified width + 0.5
> - specified height - 0.5 ≤ specified width / target ratio ≤ specified height + 0.5
> - specified height = specified width = 0
>
> The target ratio is the ratio of the [intrinsic width](https://drafts.csswg.org/css-images/#intrinsic-width) to the [intrinsic height](https://drafts.csswg.org/css-images/#intrinsic-height) in the resource. The specified width and specified height are the values of the [width](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width) and [height](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-height) attributes respectively.

This condition seems hard to meet if the `width` attribute is defined in terms of `%`.

Basically, people using `%` should instead be using CSS, and Lighthouse should be discouraging usage of this deprecated feature.

**Comment by eeeps** [comment]
> the size of the image element won't change once the image finishes downloading

If the image's intrinsic aspect ratio is different from the parent element's width ÷ whatever's in the height attribute, it may. Also % seems to break the intrinsic aspect ratios for flexible images feature. https://codepen.io/eeeps/pen/BaxbLej?editors=1111

---

Some quick HTTP Archive querying says that only 1% of `<img>`s have a `width` with "%" (31% have a `width` with a bare integer).

MDN got their language from [the spec itself](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width), which states,

> The attributes, if specified, must have values that are [valid non-negative integers](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-non-negative-integer).

In addition,

> If both attributes are specified, then one of the following statements must be true:
>
> - specified width - 0.5 ≤ specified height * target ratio ≤ specified width + 0.5
> - specified height - 0.5 ≤ specified width / target ratio ≤ specified height + 0.5
> - specified height = specified width = 0
>
> The target ratio is the ratio of the [intrinsic width](https://drafts.csswg.org/css-images/#intrinsic-width) to the [intrinsic height](https://drafts.csswg.org/css-images/#intrinsic-height) in the resource. The specified width and specified height are the values of the [width](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-width) and [height](https://html.spec.whatwg.org/multipage/embedded-content-other.html#attr-dim-height) attributes respectively.

This condition seems hard to meet if the `width` attribute is defined in terms of `%`.

Basically, people using `%` should instead be using CSS, and Lighthouse should be discouraging remaining usage of this deprecated feature.

**Comment by adamraine** [comment]
> MDN states % should not be there. Is it allowed by the HTML spec?

From my testing, `width="100%"` works exactly the same as `style="width: 100%"` in both Chrome and Firefox. For this reason, I don't think it's fair to discourage `width="100%"` *in this audit* because it is still a valid way to give `<img>` elements an explicit size before the image downloads.

> If there is a sizes attribute, isn't it used to download the image ASAP, even before any rendering?

Not sure but I don't see how it's relevant. I was pointing out how `width="100%"` forces the browser to size the `<img>` element *even if* the image is not finished downloading when the `<img>` element is being rendered.

> Basically, people using % should instead be using CSS, and Lighthouse should be discouraging remaining usage of this deprecated feature.

The feature is indeed deprecated in HTML 5, we can discuss discouraging this in a separate audit but I don't think it makes sense to do it in the unsized images audit this issue is referencing.

**Comment by eeeps** [comment]
@adamraine 
In order to reserve layout dimensions and prevent layout shifts for fluid images without distorting the intrinsic aspect ratio of the image you need to do something like:

CSS:
```
img.fluid {
  display: block;
  width: 100%;
  height: auto;
}
```

HTML:
```
<img class=fluid
  src=800x600.jpg
  width=800
  height=600
/>
```

As you say, browsers will happily size the image to 100% of the width of the container whether that 100% is in the HTML `width` attribute, or in CSS. But `%` in the HTML attribute breaks intrinsic aspect ratio calculation (uncomment `height: auto` in the CSS on my earlier codepen to see this in action), causing the `<img>` to have a height of 0 (from the CSS's `auto`, which is overriding the HTML `height=200` attribute), until it loads. When it loads we get big, jarring layout shift.

It's ~impossible to use % in the HTML attributes without either distorting the image out of its intrinsic aspect ratio, or creating a layout shift.

**Comment by adamraine** [comment]
> It's ~impossible to use % in the HTML attributes without either distorting the image out of its intrinsic aspect ratio, or creating a layout shift.

Images displayed with the wrong aspect ratio are surfaced in best practices:

<img width="789" alt="Screen Shot 2022-10-14 at 3 14 11 PM" src="https://user-images.githubusercontent.com/6752989/195951644-09d0ddf4-3aca-44f7-9a57-509cf74c221e.png">

When I correct the aspect ratio via `height: auto`, the CLS regression is caught by the unsized images audit:

<img width="772" alt="Screen Shot 2022-10-14 at 3 24 57 PM" src="https://user-images.githubusercontent.com/6752989/195952656-165cc0a5-b735-40bd-b706-ffdffe7c1cf5.png">

So Lighthouse is covering both UX regressions, it just doesn't explicitly discourage `%` in the HTML.

**Comment by eeeps** [comment]
@adamraine Do you think it would be helpful to have this or a different audit call out or discourage `%`, which is the root cause in these cases?

**Comment by connorjclark** [comment]
It seems like a good idea to suggestion this become a deprecation console message. I'd suggest filing this somewhere under the DevTools component on crbug.com  https://bugs.chromium.org/p/chromium/issues/list

**Comment by brendankenny** [comment]
@connorjclark @adamraine is this an upstream issue then? Or close?

**Comment by adamraine** [comment]
I don't think we need to file upstream about anything. There is good browser support for the % in html.

However, I have reconsidered my position on discouraging % for the unsized images audit. While it is possible for an image to be explicitly sized with a percentage, it is also possible for the image to *not* be explicitly sized. In this case it would make sense to fail the unsized images audit because of the % units. We have a couple options here:

- Don't treat % as an explicit value (like @nhoizey originally suggested)
- Sometimes treat % as an explicit value, and detect if the parent is properly sized when the image is sized with %

The latter option would probably require more work, but it's the only way to resolve this *and* keep support for the case I mentioned in https://github.com/GoogleChrome/lighthouse/issues/14449#issuecomment-1281565930.

**Comment by paulirish** [comment]
( Apparently 3 of us wrote a comment at the same time for this. Here's my overlapping take...)

First, We _could_ drop our LH unsized-images impl and depend on the native impl that supports `document-policy: unsized-media=?0`. see https://github.com/WICG/document-policy/blob/main/document-policy-explainer.md and https://wicg.github.io/document-policy/#reporting We know that the reporting API has CDP support. So.. possible to set a `document-policy-report-only` response header and slurp up the findings via the CDP events.

Now.. a few cases we discussed:

1. Percentage width + aspect ratio is fine. (And LH passes here).  👍
2. Percentage height (w/ or w/o aspect ratio) is not fine. (But LH doesn't flag this) 👎 
3. There are other 'edge' cases like ..
   - `width: min-content`. not good. and LH fails.  👍 
   - max-width + aspect-ratio is fine (i think. assuming no width value)  but LH flags. 👎
   - i'm sure there's other situations considering the various valid values for `width`, `min-width`, `max-width` and the combinations... 

We also discussed the influence of the surrounding elements. A) a fixed height parent has significance for a percent height child. B) If there's no content in the flow following an image, there's no Layout Shift. We (without adam....) decided that LH would not check either of these cases, as.. it's just a lot of complication. :)

**Comment by paulirish** [comment]
> MDN states % should not be there. Is it allowed by the HTML spec?

@brendankenny touched base with Domenic to get to the bottom of the spec issue. Domenic says its allowed in the useragent requirements (but not really in the authoring side). https://github.com/whatwg/html/issues/8589

**Comment by paulirish** [comment]
> depend on the native impl that supports `document-policy: unsized-media=?0`

Seems compelling. But we quickly looked into the impl and are not sure it has a more robust check than us..

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/media/media_element_parser_helpers.cc;l=20-21;drc=fef0781f2b4b0694ba56a0d688802c3c47f12dc9
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/geometry/length.h;l=246-249;drc=fef0781f2b4b0694ba56a0d688802c3c47f12dc9

Also this header will affect the size of unsized images. and AFAIK you can't engage it in report-only mode. AFAICT that was never implemented.  So... nevermind on this thread. :/

**Comment by DanielHeath** [comment]
Unfortunate that lighthouse flags using `aspect-ratio` + `max-width`, since that lets the image freely fit into the document without layout shift.
